launcher/modManager/chroniclesextractor.cpp

```diff
@@ -226,14 +236,22 @@ void ChroniclesExtractor::installChronicles(QStringList exe)
 		if(!createTempDir())
 			continue;
 		
-		logGlobal->info("Copying offline installer");
 		// FIXME: this is required at the moment for Android (and possibly iOS)
 		// Incoming file names are in content URI form, e.g. content://media/internal/chronicles.exe
 		// Qt can handle those like it does regular files
 		// however, innoextract fails to open such files
 		// so make a copy in directory to which vcmi always has full access and operate on it
 		QString filepath = tempDir.filePath("chr.exe");
+		logGlobal->info("Copying offline installer from '%s' to '%s'", f.toStdString(), filepath.toStdString());
+#ifdef VCMI_ANDROID
+		{
+			auto srcStr = QAndroidJniObject::fromString(f);
+			auto dstStr = QAndroidJniObject::fromString(filepath);
+			QAndroidJniObject::callStaticObjectMethod("eu/vcmi/vcmi/util/FileUtil", "copyFileFromUri", "(Ljava/lang/String;Ljava/lang/String;)V", srcStr.object<jstring>(), dstStr.object<jstring>());
```

Comment:
looks copy-pasted

Reply:
Don't see a better solution here. Suggestions?

Reply:
just extract it to a separate function (e.g. put it to Helper) and call it in both places.

```cpp
void performNativeCopy(QString src, QString dst)
{
	auto srcStr = QAndroidJniObject::fromString(f);
	auto dstStr = QAndroidJniObject::fromString(filepath);
	QAndroidJniObject::callStaticObjectMethod("eu/vcmi/vcmi/util/FileUtil", "copyFileFromUri", "(Ljava/lang/String;Ljava/lang/String;)V", srcStr.object<jstring>(), dstStr.object<jstring>());
}
```

Reply:
Ah. Misunderstood you first.

Makes sense!